jax/scipy/optimize/line_search.py

```diff
@@ -0,0 +1,464 @@
+
+    state = state._replace(failed=~(wolfe_1 & wolfe_2), k=state.k + 1)
+
+    def backtrack(state):
+      # TODO: it may make sense to only do this once on the first iteration.
+      # Moreover, can this be taken out of cond?
+      u = grad_restricted(state.a_k)
+      state = state._replace(ngev=state.ngev + 1)
+      # state = state._replace(a_k=cond(u > 0, None, lambda *x: state.a_k * rho_neg,
+      # None, lambda *x: state.a_k * rho_pos))
+      a_k = state.a_k * jnp.where(u > 0, rho_neg, rho_pos)
+      state = state._replace(a_k=a_k)
+      return state
+
+    def finish(args):
+      state, f_kp1, g_kp1 = args
+      state = state._replace(f_k=f_kp1, g_k=g_kp1)
+      return state
+
+    state = cond(state.failed, state, backtrack, (state, f_kp1, g_kp1), finish)
+
+    return state
+
+  state = while_loop(lambda state: state.failed & (state.k < max_iterations),
+                     body,
+                     state
+                     )
+
+  def maybe_update(state):
+    f_kp1, g_kp1 = restricted_func(state.a_k)
+    state = state._replace(f_k=f_kp1, g_k=g_kp1, nfev=state.nfev + 1, ngev=state.ngev + 1)
+    return state
+
+  state = cond(state.failed, state, maybe_update, state, lambda state: state)
+
+  result = LineSearchResults(failed=state.failed, nit=state.k, nfev=state.nfev, ngev=state.ngev, a_k=state.a_k,
+                             f_k=state.f_k, g_k=state.g_k)
+
+  return state
+
+
+def _cubic_interpolation(a_1, phi_1, dphi_1, a_2, phi_2, dphi_2):
+  """
+  Computes the cubic interpolation minimiser based on two previous evaluations of the
+  restricted function. It assumes that the minimum of the interpolation is in (a_1,a_2).
+
+  Args:
+      a_1: a previous step
+      phi_1: phi(a_1)
+      dphi_1: phi'(a_1)
+      a_2: a previous step not equal to a_1
+      phi_2: phi(a_2)
+      dphi_2: phi'(a_2)
+
+  Returns: a step minimising the cubic interpolant.
+
+  """
+  # return (a_1 + a_2)/2.
+  d1 = dphi_1 + dphi_2 - 3. * (phi_1 - phi_2) / (a_1 - a_2)
+  d2 = jnp.sign(a_2 - a_1) * jnp.sqrt(d1 ** 2 - dphi_1 * dphi_2)
+  a_3 = a_1 - (a_2 - a_1) * (dphi_2 + d2 - d1) / (dphi_2 - dphi_1 + 2. * d2)
+  return a_3
+
+
+def _cubicmin(a, fa, fpa, b, fb, c, fc):
+  C = fpa
+  db = b - a
+  dc = c - a
+  denom = (db * dc) ** 2 * (db - dc)
+  d1 = jnp.array([[dc ** 2, -db ** 2],
+                  [-dc ** 3, db ** 3]])
+  A, B = jnp.dot(d1, jnp.array([fb - fa - C * db, fc - fa - C * dc])) / denom
+
+  radical = B * B - 3. * A * C
+  xmin = a + (-B + jnp.sqrt(radical)) / (3. * A)
+
+  return xmin
+
+
+def _quadmin(a, fa, fpa, b, fb):
+  D = fa
+  C = fpa
+  db = b - a
+  B = (fb - D - C * db) / (db ** 2)
+  xmin = a - C / (2. * B)
+  return xmin
+
+
+def _binary_replace(replace_bit, original_dict, new_dict, keys):
+  """
+  Similar to np.where, but fewer ops.
+  """
+  not_replace_bit = ~replace_bit
+  out = dict()
+  for key in keys:
+    out[key] = not_replace_bit * original_dict[key] + replace_bit * new_dict[key]
+  return out
```

Comment:
I don't think I understand the motivation here. How is this better than using `where`?

I suspect XLA can optimize `where` to be equivalent to these sorts of optimizations.

Reply:
True, I should trust XLA more, maybe.

Reply:
If you can find any case where this is faster than `where` under JIT, please file a separate issue with a reproduction and we can pass that on a bug to the XLA team.

Reply:
I'll add this to the list of things to profile.